src/group_theory/dihedral.lean

```diff
@@ -0,0 +1,178 @@
+the rotations of the `n`-gon by `2πi/n`, and `sr i` represents the reflections of the `n`-gon.
+`dihedral 0` corresponds to the infinite dihedral group.
+-/
+
+/--
+For `n ≠ 0`, `dihedral n` represents the symmetry group of the regular `n`-gon. `r i` represents
+the rotations of the `n`-gon by `2πi/n`, and `sr i` represents the reflections of the `n`-gon.
+`dihedral 0` corresponds to the infinite dihedral group.
+-/
+@[derive decidable_eq]
+inductive dihedral (n : ℕ) : Type
+| r : zmod n → dihedral
+| sr : zmod n → dihedral
+
+namespace dihedral
+
+variables {n : ℕ}
+
+/--
+Multiplication of the dihedral group.
+-/
+def mul : dihedral n → dihedral n → dihedral n
+| (r i) (r j) := r (i + j)
+| (r i) (sr j) := sr (j - i)
+| (sr i) (r j) := sr (i + j)
+| (sr i) (sr j) := r (j - i)
+
+/--
+The identity `1` is the rotation by `0`.
+-/
+def one : dihedral n := r 0
+
+instance : inhabited (dihedral n) := ⟨one⟩
+
+/--
+The inverse of a an element of the dihedral group.
+-/
+def inv : dihedral n → dihedral n
+| (r i) := r (-i)
+| (sr i) := sr i
+
+/--
+The group structure on `dihedral n`.
+-/
+instance : group (dihedral n) :=
+{ mul := mul,
+  mul_assoc :=
+  begin
+    rintros (a | a) (b | b) (c | c);
+    simp only [mul];
+    ring,
+  end,
+  one := one,
+  one_mul :=
+  begin
+    rintros (a | a),
+    exact congr_arg r (zero_add a),
+    exact congr_arg sr (sub_zero a),
+  end,
+  mul_one := begin
+    rintros (a | a),
+    exact congr_arg r (add_zero a),
+    exact congr_arg sr (add_zero a),
+  end,
+  inv := inv,
+  mul_left_inv := begin
+    rintros (a | a),
+    exact congr_arg r (neg_add_self a),
+    exact congr_arg r (sub_self a),
+  end }
+
+lemma r_mul_r (i j : zmod n) : r i * r j = r (i + j) := rfl
+lemma r_mul_sr (i j : zmod n) : r i * sr j = sr (j - i) := rfl
+lemma sr_mul_r (i j : zmod n) : sr i * r j = sr (i + j) := rfl
+lemma sr_mul_sr (i j : zmod n) : sr i * sr j = r (j - i) := rfl
+
+lemma one_def : (1 : dihedral n) = r 0 := rfl
+
+private def fintype_helper : (zmod n ⊕ zmod n) ≃ dihedral n :=
+{ inv_fun := λ i, match i with
+                 | (r j) := sum.inl j
+                 | (sr j) := sum.inr j
+                 end,
+  to_fun := λ i, match i with
+                 | (sum.inl j) := r j
+                 | (sum.inr j) := sr j
+                 end,
+  left_inv := by rintro (x | x); refl,
+  right_inv := by rintro (x | x); refl }
+
+/--
+If `0 < n`, then `dihedral n` is a finite group.
+-/
+instance [fact (0 < n)] : fintype (dihedral n) := fintype.of_equiv _ fintype_helper
+
+instance : nontrivial (dihedral n) := ⟨⟨r 0, sr 0, dec_trivial⟩⟩
+
+/--
+If `0 < n`, then `dihedral n` has `2n` elements.
+-/
+lemma card [fact (0 < n)] : fintype.card (dihedral n) = 2 * n :=
+begin
+  rw ←fintype.card_eq.mpr ⟨fintype_helper⟩,
+  change fintype.card (zmod n ⊕ zmod n) = 2 * n,
+  rw [fintype.card_sum, zmod.card, two_mul]
+end
+
+lemma r_one_pow (k : ℕ) : (r 1 : dihedral n) ^ k = r k :=
```

Comment:
Should this be `simp`?

Reply:
I think so? From the definition of `dihedral`, both of these represent `r^k`, so I guess the form in the inductive type (ie `r k`) should be the simp normal form.